climpred/smoothing.py

```diff
@@ -0,0 +1,248 @@
+import numpy as np
+
+try:
+    import xesmf as xe
+except ImportError:
+    xe = None
+
+
+def spatial_smoothing_xesmf(
+    ds,
+    d_lon_lat_dict={'lon': 5, 'lat': 5},
+    method='bilinear',
+    periodic=False,
+    filename=None,
+    reuse_weights=True,
+):
+    """
+    Quick regridding function. Adapted from
+    https://github.com/JiaweiZhuang/xESMF/pull/27/files#diff-b537ef68c98c2ec11e64e4803fe4a113R105.
+
+    Args:
+        ds (xarray-object): Contain input and output grid coordinates.
+            Look for variables ``lon``, ``lat``, and optionally ``lon_b``,
+            ``lat_b`` for conservative method.
+            Shape can be 1D (Nlon,) and (Nlat,) for rectilinear grids,
+            or 2D (Ny, Nx) for general curvilinear grids.
+            Shape of bounds should be (N+1,) or (Ny+1, Nx+1).
+         d_lon_lat_dict (dict): optional
+            Longitude/Latitude step size (grid resolution); if not provided,
+            lon will equal 5 and lat will equal lon
+            (optional)
+         method (str): Regridding method. Options are:
+            - 'bilinear'
+            - 'conservative', **need grid corner information**
+            - 'patch'
+            - 'nearest_s2d'
+            - 'nearest_d2s'
+         periodic (bool): Periodic in longitude? Default to False. optional
+            Only useful for global grids with non-conservative regridding.
+            Will be forced to False for conservative regridding.
+         filename (str): Name for the weight file. (optional)
+            The default naming scheme is:
+                 {method}_{Ny_in}x{Nx_in}_{Ny_out}x{Nx_out}.nc
+                 e.g. bilinear_400x600_300x400.nc
+         reuse_weights (bool) Whether to read existing weight file to save
+            computing time. False by default. (optional)
+
+        Returns:
+            ds (xarray.object) regridded
+    """
+
+    if xe is None:
+        raise ImportError(
+            'xesmf is not installed; see'
+            'https://xesmf.readthedocs.io/en/latest/installation.html'
+        )
+
+    def _regrid_it(da, d_lon, d_lat, **kwargs):
+        """
+        Global 2D rectilinear grid centers and bounds
+
+        Args:
+            da (xarray.DataArray): Contain input and output grid coords.
+                Look for variables ``lon``, ``lat``, ``lon_b``, ``lat_b`` for
+                conservative method, and ``TLAT``, ``TLON`` for CESM POP grid
+                Shape can be 1D (Nlon,) and (Nlat,) for rectilinear grids,
+                or 2D (Ny, Nx) for general curvilinear grids.
+                Shape of bounds should be (N+1,) or (Ny+1, Nx+1).
+            d_lon (float): Longitude step size, i.e. grid resolution
+            d_lat (float): Latitude step size, i.e. grid resolution
+        Returns:
+            da : xarray DataArray with coordinate values
+        """
+
+        def check_lon_lat_present(da):
```

Comment:
A function nested in a function nested in a function is preeetty uncommon and not sure if it's a good idea